---
output: rmarkdown::github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment="##", fig.retina=2, fig.path = "README_figs/README-")
```

#**README**
##**R code for: Parras-Moltó & Aguirre de Cárcer, 2019**

This document describes the use of FunCongr.R, as presented in "Assessment of phylo-functional coherence along the bacterial phylogeny and taxonomy"" by Parras-Moltó and Aguirre de Cárcer (https://www.biorxiv.org/content/10.1101/795914v1).The procedure employed trees, taxonomic assignments, and 16S rRNA gene sequences from GTDB (https://gtdb.ecogenomic.org), as well as the full list of genome annotations from proGenomes (http://progenomes.embl.de/).

```
`r version[['version.string']]`
```
###**FunCongr.R**

###**Dependencies**

**Packages**

· `ape` v`r packageVersion("ape")`. Analyses of phylogenetics and evolution

· `tree` v`r packageVersion("tree")`. Fit a classification or regression tree

· `gtools` v`r packageVersion("gtools")`. Various R programming tools

· `phangorn` v`r packageVersion("phangorn")`. Phylogenetic analysis in R

**Files located in the same folder as the script**

1- gtdb_r86.ssu.bacteria.fasttree_name_pruned.tree

2- tab_info_pair.txt

```{r,echo=FALSE,warning=FALSE}
library("ape")
library("tree")
library("gtools")
library("phangorn") #Packages needed for the analysis

#Functions

isEmpty <- function(x) { #This function checks if a data frame is empty or not
  return(length(x)==0)
}
```



Before the use of the script itself we prepared the input data; First we transformed genome annotations into a binary presence/absence table of gene content per genome (COG annotations were employed for consistency, representing >77% of the total annotations). 

```{r}
tab_pres=read.table("info_cogs.txt",header=T)
head(tab_pres[,1:8])
```

Then, for each pairwise comparison between genomes we produced 16S rRNA gene distances and Jaccard distances based on gene content (average of A-B and B-A), summarized in a four columns table (tab_info_pair.txt).Constant node labels were given to the input tree using makeNodeLabel from ape package in R (_name). Subsequently, the tree was pruned to contain only leaves with annotated genomes (_name_pruned).

```{r}
tree=read.tree("gtdb_r86.ssu.bacteria.fasttree_name_pruned.tree") #The tree is loaded here
tab_info_pair=read.table("tab_info_pair.txt") #Within this table we have all the distances values (Jaccard and 16S) for each  posible pair.
tab_info_pair$comb<-paste(tab_info_pair$V1,tab_info_pair$V2,sep = "")

head(tab_info_pair[,1:5])
```

FunCongr.R reads the named_pruned tree, as well as the four columns table in the same directory, and prints out, fore each evaluated node, node name and hour (nodes passing the test) plus within-node average and standard deviation of Jaccard values (nodes failing the test, e.g. output_16s.log). The script does not evaluate nodes with fewer than 5 members or nodes presenting descendant nodes that already failed the test. Due to its nature, the script runs in a single processor, which, due to the size of the tree, translated into a lengthy (>400h) run.

The output file is a list of so called "bad nodes". It is checked if our nodes passes the threeshold of the ecdf function. In case the p-value obtained is greater of 0.05, the node is considerated as "bad" and its leaves are saved to avoid the evaluation of the subsequents nodes that posseses the sames leaves.

```{r}
node_list=paste(unlist(read.table("bad_nodes.txt")))
print(node_list)
```

###**Paint_tree.R**

The results are later parsed with Paint_tree.R, which reads in the same directory the same tree and four columns table, as well as the metadata table containing the taxonomic affiliation of each leave (bac_metadata_r86.tsv) and a list of failing nodes identified with FunCongr.R (e.g. bad_nodes.txt). The scripts follows the rule that non-functionally coherent nodes are those that either failed the test or present a descendant node that failed the test. Then, for each last node were functional congruence was observed along each branching path it outputs a table with node ID, node label, number of leaves, and average plus standard deviation values of within-node 16S rRNA gene and Jaccard-based distances, and the node's 80% consensus taxonomy) (nodes_info.txt), as well as a painted tree as a pdf image. 

###**Dependencies**

**Packages**

· `ape` v`r packageVersion("ape")`. Analyses of phylogenetics and evolution

· `tree` v`r packageVersion("tree")`. Fit a classification or regression tree

· `phangorn` v`r packageVersion("phangorn")`. Phylogenetic analysis in R

· `phytools` v`r packageVersion("phytools")`. Phylogenetic tools for comparative biology (and other things)

· `ggtree` v`r packageVersion("ggtree")`. Visualizing phylogenetic tree

· `ggplot2` v`r packageVersion("ggplot2")`. Create elegant data visualisations using the grammar of graphics

· `randomcoloR` v`r packageVersion("randomcoloR")`. Get a pretty random color

· `tidygraph` v`r packageVersion("tidygraph")`. A tidy API for graph manipulation

**Files located in the same folder as the script**

1- gtdb_r86.ssu.bacteria.fasttree_name_pruned.tree

2- tab_info_pair.txt

3- bac_metadata_r86.tsv(https://data.gtdb.ecogenomic.org/releases/release86/86.0/bac_metadata_r86.tsv)

4- bad_nodes.txt



**Resulting plot**

```{r,echo=FALSE,message=FALSE,warning=FALSE}
library("ape")
library("phytools")
library("tree")
library("phangorn")
library("ggtree")
library("ggplot2")
library("randomcoloR")
library("tidygraph") #Packages needed for the analysis

Element_Exists_Check = function( full_index_path ){ #We will use this function to check if a certain node exists, avoiding an error
  tryCatch({
    len_element = length(full_index_path)
    exists_indicator = ifelse(len_element > 0, T, F)
    return(exists_indicator)
  }, error = function(e) {
    return(F)
  })
}
isEmpty <- function(x) { #This function checks if a data frame is empty or not
  return(length(x)==0)
}
Nodes_info = function( node,taxono,p4,data_out_res){ #This function gets the node name and prints the information for each node: number of leaves, jaccard and 16S mean and taxonomy
  
  mini=extract.clade(input_tree,as.character(node)) #We extract the subtree for each node, to get their leaves
  taxa_data<-data.frame(matrix(ncol=2))
  taxa_data[1,1]<-"dummy"
  taxa_data[1,2]<-0
  taxa_data2<-data.frame(matrix(ncol=3,nrow=7))
  taxa_data2[,1]<-c("d","p","c","o","f","g","s")
  taxa_data2[,2]<-0
  concat_taxo=NULL
  cont=2
  
  ### In this block, we get the taxonomy for each leave, save them and compute if each taxonomic range is represented at least by the 80% of the leaves
  
  for(o in 1:length(mini$tip.label)) 
  {
    taxa_each_one<-taxono[mini$tip.label[o]==taxono[,1],92]
    
    node_taxonomy_each<-strsplit(taxa_each_one,split = ";")
    
    for(e in 1:length(node_taxonomy_each[[1]]))
    {
      if(isEmpty(taxa_data[taxa_data[,1]==node_taxonomy_each[[1]][e],2]))
      {
        taxa_data[cont,1]<-node_taxonomy_each[[1]][e]
        taxa_data[cont,2]<-1
        cont=cont+1
      }else
      {
        taxa_data[taxa_data[,1]==node_taxonomy_each[[1]][e],2]<-taxa_data[taxa_data[,1]==node_taxonomy_each[[1]][e],2]+1
      }
    }
  }
  
  taxa_data<-taxa_data[-1,]
  
  for(e in 1:dim(taxa_data)[1])
  {
    node_taxonomy_each<-(strsplit(taxa_data[e,1],split = "_"))[[1]]
    
    if(taxa_data2[taxa_data2[,1]==node_taxonomy_each[1],2]<taxa_data[e,2])
    {
      taxa_data2[taxa_data2[,1]==node_taxonomy_each[1],2]=taxa_data[e,2]
      taxa_data2[taxa_data2[,1]==node_taxonomy_each[1],3]=node_taxonomy_each[3]
    }
  }
  
  taxa_data2[,2]<-taxa_data2[,2]/length(mini$tip.label)
  b_impr=0
  
  for(e in 1:dim(taxa_data2)[1])
  {
    if(taxa_data2[e,2]>0.8 & b_impr==0)
    {
      concat_taxo=paste(concat_taxo,";",taxa_data2[e,1],"__",taxa_data2[e,3],sep="")
    }else
    {
      b_impr=1
    }
  }
  
  ###Here we compute the jaccard and 16S mean for each node, as those values weren't saved properly by FunCongr.R
  
  combinations_leaves=data.frame(expand.grid(mini$tip.label,mini$tip.label))
  
  combinations_leaves$Var1=as.character(combinations_leaves$Var1)
  combinations_leaves$Var2=as.character(combinations_leaves$Var2)
  
  values_16S=NULL
  
  values_j=NULL
  
  combinations_parse<-paste(combinations_leaves$Var1,combinations_leaves$Var2,sep = "")
  values_j<-tab_info_pair[tab_info_pair$comb %in% combinations_parse,4]
  values_16S<-tab_info_pair[tab_info_pair$comb %in% combinations_parse,3]
  
  combinations_parse<-paste(combinations_leaves$Var2,combinations_leaves$Var1,sep = "")
  values_j<-c(values_j,tab_info_pair[tab_info_pair$comb %in% combinations_parse,4])
  values_16S<-c(values_16S,tab_info_pair[tab_info_pair$comb %in% combinations_parse,3])
  
  ##The info is saved into a list, and returned as a tree and data frame
  
  #p4<-p4+
  #  geom_cladelabel(edge_table[edge_table$par.name==node,1][1], paste(node,". #Leaves: ",length(mini$tip.label),". Jaccard: ",round(mean(values_j),digits = 3),"Â±",round(sd(values_j),digits = 3),". 16S: ",round(mean(values_16S),digits = 3),"Â±",round(sd(values_16S),digits = 3),". Taxonomy: ",concat_taxo,sep = ''), barsize=2, offset.text=1.5, hjust=0.5, fontsize=3)
  
  data_out_res<-rbind(c(node,edge_table[edge_table$par.name==node,1][1],length(mini$tip.label),round(mean(values_j),digits = 3),round(sd(values_j),digits = 3),round(mean(values_16S),digits = 3),round(sd(values_16S),digits = 3),concat_taxo),data_out_res)
  
  return(list("tree"=p4,"data"=data_out_res))
}

tree=read.tree("gtdb_r86.ssu.bacteria.fasttree_name_pruned.tree") #Here we read the tree
tab_info_pair=read.table("tab_info_pair.txt") #Within this table we have all the distances values (Jaccard and 16S) for each  posible pair
tab_info_pair$comb<-paste(tab_info_pair$V1,tab_info_pair$V2,sep = "")
require(data.table)
taxono=as.data.frame(fread("bac_metadata_r86.tsv")) #The information of the taxonomy is loaded from this metadata table
data_out_res=data.frame(matrix(ncol=8))
colnames(data_out_res)=c("Node","ID","Leaves","Jaccard mean","Jaccard sd","16S mean","16S sd","Taxonomy")
colores_aleat<-c("#4af7b2","#f9c0b6","#e266b3","#692ed1","#a6fcce","#9597e2","#7882cc","#6b8c00","#c91f02","#bddcfc","#f29b8a","#515fcc","#554aba","#c987e5","#6feda1","#d67564","#d8d84e","#d1aa40","#92f464","#83f7c8","#f49e9c","#83ea94","#ffc966","#ffaae1","#e9ff89","#f2c59d","#99eafc","#95c2d8","#f9b8e0","#ea0edb","#cc6f1e","#def79e","#91e50b","#50e21b","#d64fa6","#afa7f2","#b6fca6","#ee4cf7","#72fff0","#32d164","#7aefe1","#2634af","#386ca0","#79cff7","#c0d838","#a9f77b","#87d7e5","#5bf21a","#ff0213","#c088d8","#f47a9b","#968fd6","#c9d7ff","#efc970","#e05be5","#0296e0","#bce866","#53a81e","#bbbcf7","#17e867","#f2214b","#0dc609","#d63960")
cont_col_rand<-1
input_tree<-tree

####With this function we are able to get all the information from the nodes and leaves, their ID and given name

node_labels_in_edge <- input_tree$node.label[input_tree$edge[,1]-Ntip(input_tree)]
tips_nodes <- input_tree$edge[,2]

select.tip.or.node <- function(element, tree) {
  ifelse(element < Ntip(tree)+1, tree$tip.label[element], tree$node.label[element-Ntip(tree)])
}

edge_table <- data.frame(
  "parent" = input_tree$edge[,1],
  "par.name" = sapply(input_tree$edge[,1], select.tip.or.node, tree = input_tree),
  "child" = input_tree$edge[,2],
  "chi.name" = sapply(input_tree$edge[,2], select.tip.or.node, tree = input_tree)
)

edge_table$parent=as.numeric(edge_table$parent)
edge_table$child=as.numeric(edge_table$child)
edge_table$par.name=as.character(edge_table$par.name)
edge_table$chi.name=as.character(edge_table$chi.name)

####

node_list=paste(unlist(read.table("bad_nodes.txt"))) #In this table we saved the info of the nodes considerated as "bad nodes" after the FunCogr.R script
painted_leaves=NULL #We will save the leaves that are already painted on the tree to avoid reprinting them

for (i in 1:length(node_list)) #Then, we print the info of all descendant nodes from all the "bad nodes" and save their leaves
{
  descendant_nodes=Descendants(input_tree,edge_table[edge_table$par.name==node_list[i],1],type = "child")[[1]]
  
  painted_leaves=c(painted_leaves,paste(unlist(Descendants(input_tree,descendant_nodes[1]))),paste(unlist(Descendants(input_tree,descendant_nodes[2]))))
}

p4 <- ggtree(tree,layout="circular") + #First, we print the tree with the ggtree function
  #p4 <- ggtree(tree) +
  geom_tiplab(size=0)

for (i in 1:length(node_list)) #Then, we print the info of all descendant nodes from all the "bad nodes" and save their leaves
{
  descendant_nodes=Descendants(input_tree,edge_table[edge_table$par.name==node_list[i],1],type = "child")[[1]]
  
  painted_leaves=c(painted_leaves,paste(unlist(Descendants(input_tree,descendant_nodes[1]))),paste(unlist(Descendants(input_tree,descendant_nodes[2]))))
  
  if(length(unlist(Descendants(input_tree,descendant_nodes[1])))>4) #We print squares over each descendant node and print the info for each one (number of leaves, means, sds and taxonomy)
  {
    p4<-p4+
      geom_hilight(node=descendant_nodes[1], fill=colores_aleat[cont_col_rand], alpha=0.5)
    cont_col_rand<-cont_col_rand+1
    
    if(!is.na(edge_table[edge_table$parent==descendant_nodes[1],2][1]))
    {
      nodes_out<-Nodes_info(edge_table[edge_table$parent==descendant_nodes[1],2][1],taxono,p4,data_out_res) #We call the Nodes_info function to print the information of each "good node"
      p4<-nodes_out$tree #We save the modified info of the tree
      data_out_res<-nodes_out$data #We save the info of the modified data frame with the results, too
    }
  }
  
  if(length(unlist(Descendants(input_tree,descendant_nodes[2])))>4) #Here we do the same as before for the other descendant node
  {
    p4<-p4+
      geom_hilight(node=descendant_nodes[2], fill=colores_aleat[cont_col_rand], alpha=0.5)
    cont_col_rand<-cont_col_rand+1
    
    if(!is.na(edge_table[edge_table$parent==descendant_nodes[2],2][1]))
    {
      nodes_out<-Nodes_info(edge_table[edge_table$parent==descendant_nodes[2],2][1],taxono,p4,data_out_res)
      p4<-nodes_out$tree
      data_out_res<-nodes_out$data
    }
  }
  
  ancest_node=Ancestors(input_tree,edge_table[edge_table$par.name==node_list[i],1])[[1]] #A list of ancestors is obtained. Those nodes are considered "bad nodes" too and won't be printed
  
  sibling_node=Siblings(input_tree,edge_table[edge_table$par.name==node_list[i],1])[[1]] #We get a list of siblings too, as they would be considered "good nodes" by defintion
  
  leaves=paste(unlist(Descendants(input_tree,sibling_node))) #A list of total leaves for each node is obtained
  
  if(length(na.omit(match(leaves,painted_leaves)))==0 & length(leaves)>4) #Here we check if the leaves of the node ares painted yet
  {
    p4<-p4+
      geom_hilight(node=sibling_node, fill=colores_aleat[cont_col_rand], alpha=0.5) #We paint the sibling node
    cont_col_rand<-cont_col_rand+1
    painted_leaves=c(painted_leaves,leaves)
    
    if(!is.na(edge_table[edge_table$parent==sibling_node,2][1])) #Here we get the information for this node too, as done before
    {
      nodes_out<-Nodes_info(edge_table[edge_table$parent==sibling_node,2][1],taxono,p4,data_out_res)
      p4<-nodes_out$tree
      data_out_res<-nodes_out$data
    }
  }
  
  for(o in 1:length(ancest_node)) #For each ancestor node we paint their siblings, in case their leaves aren't painted yet
  {
    if(Element_Exists_Check(Siblings(input_tree, ancest_node[o])[[1]]))
    {
      sibling_node=Siblings(input_tree,ancest_node[o])[[1]]
      
      leaves=paste(unlist(Descendants(input_tree,sibling_node)))
      
      if(length(na.omit(match(leaves,painted_leaves)))==0 & length(leaves)>4)
      {
        p4<-p4+
          geom_hilight(node=sibling_node, fill=colores_aleat[cont_col_rand], alpha=0.5)
        cont_col_rand<-cont_col_rand+1
        painted_leaves=c(painted_leaves,leaves)
        
        if(!is.na(edge_table[edge_table$parent==sibling_node,2][1]))
        {
          nodes_out<-Nodes_info(edge_table[edge_table$parent==sibling_node,2][1],taxono,p4,data_out_res)
          p4<-nodes_out$tree
          data_out_res<-nodes_out$data
        }
      }
    }
  }
  
}


#p4<-p4+geom_tiplab(aes(subset=node==edge_table[edge_table$par.name==node_list[1],1], label=node_list[1], parse=T)) + #With this lines we paint the "bad nodes" on their position on the tree
#  geom_tiplab(aes(subset=node==edge_table[edge_table$par.name==node_list[2],1], label=node_list[2], parse=T)) +
 # geom_tiplab(aes(subset=node==edge_table[edge_table$par.name==node_list[3],1], label=node_list[3], parse=T)) +
  #geom_tiplab(aes(subset=node==edge_table[edge_table$par.name==node_list[4],1], label=node_list[4], parse=T)) +
  #geom_tiplab(aes(subset=node==edge_table[edge_table$par.name==node_list[5],1], label=node_list[5], parse=T))

#pdf("tree_paint_jj_circ.pdf",height = 20, width = 20) #All information is printed and saved
#pdf("tree_paint_8.pdf",height = 200, width = 20)
plot(p4)
#dev.off()

data_out_res=data_out_res[-dim(data_out_res)[1],]

#write.table(data_out_res,"nodes_info.txt",sep="\t",row.names = F,col.names = T)
```

Resulting nodes info are saved to "node_info.txt" file, with the: "Node","ID","Leaves","Jaccard mean","Jaccard sd","16S mean","16S sd" and "Taxonomy" information fields.

```{r}
head(data_out_res)
```

